pkg/queue/concurrency_state.go

```diff
@@ -0,0 +1,113 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package queue
+
+import (
+	"net/http"
+
+	"go.uber.org/zap"
+)
+
+// ConcurrencyStateHandler tracks the in flight requests for the pod. When the requests
+// drop to zero, it runs the `pause` function, and when requests scale up from zero, it
+// runs the `resume` function. If either of `pause` or `resume` are not passed, it runs
+// the respective local function(s). The local functions are the expected behavior; the
+// function parameters are enabled primarily for testing purposes.
+func ConcurrencyStateHandler(endpoint string, h http.Handler, logger *zap.SugaredLogger, pause, resume func()) http.HandlerFunc {
+	logger.Info("Concurrency state tracking enabled")
+
+	if pause == nil {
+		pause = func() {
+			logger.Info("Requests dropped to zero, pausing pod ...")
+			pauseEndpoint := endpoint + ":9696/freeze"
+			req, err := http.NewRequest(http.MethodPost, pauseEndpoint, nil)
+			if err != nil {
+				panic(err)
+			}
+			req.Header.Add("Token", "") // TODO: use serviceaccountToken from projected volume
+			resp, err := http.DefaultClient.Do(req)
+			if err != nil {
+				panic(err)
```

Comment:
I think we need to think through these failure cases, a panic doesn't seem good enough here.

Reply:
Thinking out loud a bit here...

If the resume call fails, in an ideal state I think I'd want to log the failure and then retry at least one more time (perhaps it was just due to a network blip?). If still unable to resume the container, then I think I'd want to launch another pod and send the request to it (though not sure how feasible that would be). The key here is to get an active pod back up.

If the case where the pause call fails, I think that logging the failure and retrying might be sufficient. A not-paused paused container doesn't really negatively impact the end-user (their service would've scaled down anyways). That said, an alternative might be to kill the pod and respawn it in a paused state (which might be a better choice from a provider perspective). 

@julz what's your take?

Reply:
my instinct is let's spin out an issue for retrying, maybe another for more generally thinking through what should happen here (e.g. should we fail readiness/liveness probes maybe so that k8s routes around us/kills us and brings up another container?), and then for right now land this with a panic (since it's an experimental branch throwing our hands in the air is probably better than logging, and recovering needs more thought - but I would like an issue to exist for at _least_ retrying).

@markusthoemmes any thoughts?

Reply:
(I've reverted to the panic, but kept the log message as I think it's still useful info to have)